### PR TITLE
#124: introducing macroParameters for generators

### LINKED ISSUE #124: Introducing macroParameterSet for generators
it follows dynawo issue 1459

https://github.com/dynawo/dynawo/issues/1459

### PR DESCRIPTION
**Pull request by marcochiaramello**
Closes #124

### CODE REVIEW

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Par.cpp [hunk 1]
```diff
@@ -53,6 +53,120 @@ buildReference(const std::string& name, const std::string& origName, const std::
   return ref;
 }
 
+static constexpr double kGoverNullValue_ = 0.;     ///< KGover null value
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:56]: needed in helper class

File: sources/Outputs/src/Par.cpp [hunk 2]
```diff
@@ -62,29 +176,41 @@ Par::write() {
...
+    }
+    if (it->isUsingDiagram()) {
+      dynamicModelsToConnect->addParametersSet(writeGenerator(*it, def_.basename, def_.dirname));
+    } else {
+      if(!dynamicModelsToConnect->hasParametersSet(helper::getGeneratorParameterSetId(it->model, DYN::doubleIsZero(it->targetP)))) {
+        dynamicModelsToConnect->addParametersSet(writeConstantGeneratorsSets(def_.activePowerCompensation, it->model, DYN::doubleIsZero(it->targetP))); 
+      }
     }
   }
+  // loop on hvdcs
```
Comment (rosiereflo) [inline sources/Outputs/src/Par.cpp:198]: useless comment

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/include/Constants.h [hunk 3]
```diff
@@ -57,7 +57,9 @@ const std::string signalNGeneratorFixedPParId{"signalNGeneratorFixedP"};
 const std::string impSignalNGeneratorParId{"impSignalNGenerator"};              ///< PAR id for generators using signal N with impendance
 const std::string impSignalNGeneratorFixedPParId{"impSignalNGeneratorFixedP"};  ///< PAR id for generators using signal N with impendance with fixed P
 const std::string propSignalNGeneratorParId{"propSignalNGenerator"};      ///< PAR id for generators using prop signal N
+const std::string propSignalNGeneratorFixedPParId{"propSignalNGeneratorFixedP"};      ///< PAR id for generators using prop signal N with fixed P
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Outputs/include/Constants.h:60]: done, thanks

File: sources/Outputs/include/Par.h [hunk 4]
```diff
@@ -102,35 +112,33 @@ class Par {
    * @param fixedP boolean to determine if the set represents a generator with a targetP equal to 0
    */
   static boost::shared_ptr<parameters::ParametersSet>
-  updateSignalNGenerator(const std::string& modelId, dfl::inputs::Configuration::ActivePowerCompensation activePowerCompensation, bool fixedP);
+  updateSignalNGenerator(const std::string& modelId, const dfl::inputs::Configuration::ActivePowerCompensation& activePowerCompensation, bool fixedP);
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Outputs/include/Par.h:115]: solved

File: sources/Outputs/src/Par.cpp [hunk 5]
```diff
@@ -53,6 +53,120 @@ buildReference(const std::string& name, const std::string& origName, const std::
...
 }
 
+static constexpr double kGoverNullValue_ = 0.;     ///< KGover null value
+static constexpr double kGoverDefaultValue_ = 1.;  ///< KGover default value
+
+static std::string getMacroParameterSetId(const std::string& modelType) {
+  return "macro_" + modelType;
+}
+
+static std::string getMacroParameterSetId(const algo::GeneratorDefinition::ModelType& modelType, bool fixedP) {
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:63]: done

File: sources/Outputs/src/Par.cpp [hunk 6]
```diff
@@ -62,29 +176,41 @@ Par::write() {
...
+      if(!dynamicModelsToConnect->hasParametersSet(helper::getGeneratorParameterSetId(it->model, DYN::doubleIsZero(it->targetP)))) {
+        dynamicModelsToConnect->addParametersSet(writeConstantGeneratorsSets(def_.activePowerCompensation, it->model, DYN::doubleIsZero(it->targetP))); 
+      }
     }
   }
+  // loop on hvdcs
   for (auto it = def_.hvdcLines.begin(); it != def_.hvdcLines.end(); ++it) {
     dynamicModelsToConnect->addParametersSet(writeHdvcLine(it->second));
   }
+  // adding parameters sets related with remote voltgae control or multple generator regulating same bus
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:202]: thx

File: sources/Outputs/src/Par.cpp [hunk 7]
```diff
@@ -62,29 +176,41 @@ Par::write() {
...
-    auto set = writeGenerator(*it, def_.basename, def_.dirname, def_.activePowerCompensation);
-    if (set) {
-      dynamicModelsToConnect->addParametersSet(set);
+    if(!dynamicModelsToConnect->hasMacroParametersSet(helper::getMacroParameterSetId(it->model, DYN::doubleIsZero(it->targetP))) && it->isUsingDiagram()) {
+      dynamicModelsToConnect->addMacroParameterSet(helper::buildMacroParameterSet(it->model, def_.activePowerCompensation, DYN::doubleIsZero(it->targetP)));
+    }
+    if (it->isUsingDiagram()) {
+      dynamicModelsToConnect->addParametersSet(writeGenerator(*it, def_.basename, def_.dirname));
+    } else {
+      if(!dynamicModelsToConnect->hasParametersSet(helper::getGeneratorParameterSetId(it->model, DYN::doubleIsZero(it->targetP)))) {
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:193]: comments have been added, thanks

File: sources/Outputs/src/Par.cpp [hunk 8]
```diff
@@ -195,22 +197,22 @@ Par::write() {
       }
     }
   }
-  // loop on hvdcs
   for (auto it = def_.hvdcLines.begin(); it != def_.hvdcLines.end(); ++it) {
     dynamicModelsToConnect->addParametersSet(writeHdvcLine(it->second));
   }
-  // adding parameters sets related with remote voltgae control or multple generator regulating same bus
+  // adding parameters sets related with remote voltage control or multple generator regulating same bus
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Par.cpp:203]: typo "multple"

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Par.cpp [hunk 9]
```diff
@@ -339,8 +341,8 @@ Par::writeGenerator(const algo::GeneratorDefinition& def, const std::string& bas
 
   //  Use the hash id in exported files to prevent use of non-ascii characters
   auto set = boost::shared_ptr<parameters::ParametersSet>(new parameters::ParametersSet(hashIdStr));
-  if(def.isUsingDiagram())
-    set->addMacroParSet(boost::shared_ptr<parameters::MacroParSet>(new parameters::MacroParSet(helper::getMacroParameterSetId(def.model, DYN::doubleIsZero(def.targetP)))));
+  // The macroParSet is associated to a macroParameterSet via the id
+  set->addMacroParSet(boost::shared_ptr<parameters::MacroParSet>(new parameters::MacroParSet(helper::getMacroParameterSetId(def.model, DYN::doubleIsZero(def.targetP)))));
```
Comment (lecourtoisflo) [inline sources/Outputs/src/Par.cpp:345]: Why does the condition on isUsingDiagram been removed ?

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Par.cpp (hunk 9 above)
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:345]: the check is already implemented (line 193) when calling the method writeGenerator, so this second condition is useless

File: sources/Outputs/src/Par.cpp (hunk 8 above)
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:203]: thanks!

**Review by lecourtoisflo — APPROVED** [review]

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**